src/service/viewport/viewport-impl.js

```diff
@@ -425,20 +433,38 @@ export class ViewportImpl {
 
   /** @override */
   animateScrollIntoView(element, pos = 'top', opt_duration, opt_curve) {
-    devAssert(
-      !opt_curve || opt_duration !== undefined,
-      "Curve without duration doesn't make sense."
-    );
+    if (IS_SXG) {
+      const {promise, resolve} = new Deferred();
+      const waiter = debounce(
+        this.ampdoc.win,
+        () => {
+          this.ampdoc.win.removeEventListener('scroll', waiter);
+          resolve();
+        },
+        300
+      );
+      this.ampdoc.win.addEventListener('scroll', waiter);
```

Comment:
Can you clarify to me (and also document in a comment above) how debouncer works here? Do I understand this right that "promise is resolved 300ms after the last scroll event"?

Reply:
That is correct. After the scroll event occurs, we wait for 300ms to call the resolve.

Reply:
I feel like something is missing here. I think you need to still call `waiter()` manually? Did you test with the case where no actual scrolling will occur and no `scroll` events will arrive? This would happen, e.g. on a page that doesn't have enough content to scroll. This is important. If we don't handle it right, we will have a never-resolving promise hanging.

Reply:
So I just tested this with an empty AMP page (which will not have any content) that does nothing but "scroll" back to the top with a "Back to Top" button (which will trigger the code inside the `scrollIntoView` function, but I moved the code in `animateScrollToView` inside that function) and checked the console for uncaught promises. I did not see any errors in the console.

Reply:
I wouldn't expect any uncaught promises in console. The problem that I expect (or rather where I see this code not quite stable) is if the promise never resolves: either successfully or with a failure.

Think of it this way. The following promise will never resolve:

```
return new Promise(resolve => {});
```

That means that if a code calls `animateScrollIntoView()` will never resolve:

```
viewport.animateScrollIntoView(element).then(() => {
  // THIS code will never execute and neither would any code that further depends on this code.
})
```

Reply:
This is also a reason why I really dislike `Deferred` class since it hides this problem. So I'm also asking you to remove `Deffered` from here and use the straightforward native `new Promise()`.

Reply:
I just tested this, and yes, the promise never resolved.

Reply:
I think that's good enough, yes. There's a way to make it more precise, but at this point I think the value is not worth it.